### PR TITLE
refactor: core: Extracting ICollateral get/set and onlyCollateral modifier

### PR DESCRIPTION
**Pull request by davidprepo**
Parent PR: https://github.com/prepo-io/prepo-monorepo/pull/653

Spect: https://www.notion.so/prepo/SC-Backlog-Tasks-0c6725bb870f4c22b7ba771347333662#2807d6e4692f4f08996261da4069097d

`IOnlyAllowCollateralCaller` was created in parent PR to keep needed parts from now-deleted `ICollateralHook`.

This PR replaces IOnlyAllowCollateralCaller with UsesCollateral, which implements `set`/`get` and `modifier` in the current implementation style for other modules.

This allows for reduced code duplication, as before this PR all contracts implemented their own getters, setters, and modifiers which were all identical.

This should most likely be moved to `prepo-shared-contracts`, but that would also require movement of `ICollateral`. This can come in a followup PR.

### CODE REVIEW

File: apps/smart-contracts/core/contracts/DepositHook.sol [hunk 1]
```diff
@@ -15,16 +15,15 @@ import "@openzeppelin/contracts/utils/structs/EnumerableMap.sol";
 
 contract DepositHook is
   IHook,
-  IOnlyAllowCollateralCaller,
   IDepositHook,
   INFTAccessHook,
   AllowlistHook,
   TokenSenderCaller,
+  UsesCollateral,
```
Comment (ramenforbreakfast) [inline apps/smart-contracts/core/contracts/DepositHook.sol:22]:
```suggestion
AllowedCollateralCaller
```
 As per previous PR, I think this should be the name to maintain consistency

File: apps/smart-contracts/core/test/DepositHook.test.ts [hunk 2]
```diff
@@ -349,39 +349,12 @@ describe('=> DepositHook', () => {
...
-
-    it('emits CollateralChange', async () => {
-      const tx = await depositHook.connect(deployer).setCollateral(collateral.address)
+    it('succeeds if role holder', async () => {
+      expect(
+        await depositHook.hasRole(await depositHook.SET_COLLATERAL_ROLE(), deployer.address)
+      ).to.eq(true)
 
-      await expect(tx).to.emit(depositHook, 'CollateralChange').withArgs(collateral.address)
+      await expect(depositHook.connect(deployer).setCollateral(collateral.address))
```
Comment (ramenforbreakfast) [inline apps/smart-contracts/core/test/DepositHook.test.ts:357]:
```suggestion
      await expect(depositHook.connect(deployer).setCollateral(collateral.address))
```
Are expects even necessary in these tests? I think we can just call them no? And if they revert then the testcase fails.

File: apps/smart-contracts/core/contracts/DepositHook.sol (hunk 1 above)
Comment (XavierEkkel) [reply apps/smart-contracts/core/contracts/DepositHook.sol:22]: yes, agree with the naming `AllowedCollateralCaller` (to match the interface)

File: apps/smart-contracts/core/contracts/UsesCollateral.sol [hunk 3]
```diff
@@ -0,0 +1,23 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity =0.8.7;
+
+import "./interfaces/ICollateral.sol";
+import "./interfaces/IUsesCollateral.sol";
+
+contract UsesCollateral is IUsesCollateral {
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/contracts/UsesCollateral.sol:7]:
```suggestion
contract AllowedCollateralCaller is IAllowedCollateralCaller {
```

File: apps/smart-contracts/core/contracts/UsesCollateral.sol [hunk 4]
```diff
@@ -0,0 +1,23 @@
+// SPDX-License-Identifier: UNLICENSED
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/contracts/UsesCollateral.sol:1]:
```suggestion
// SPDX-License-Identifier: AGPL-3.0
```

File: apps/smart-contracts/core/contracts/interfaces/IUsesCollateral.sol [hunk 5]
```diff
@@ -5,7 +5,7 @@ import "./ICollateral.sol";
 
 /// @notice Used for adding additional checks and/or data recording when
 /// interacting with the Collateral vault.
-interface IOnlyAllowCollateralCaller {
+interface IUsesCollateral {
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/contracts/interfaces/IUsesCollateral.sol:8]:
```suggestion
interface IAllowedCollateralCaller {
```

File: apps/smart-contracts/core/test/fixtures/UsesCollateralFixture.ts [hunk 6]
```diff
@@ -0,0 +1,7 @@
+import { ethers } from 'hardhat'
+import { UsesCollateral } from '../../typechain'
+
+export async function usesCollateralFixture(): Promise<UsesCollateral> {
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/fixtures/UsesCollateralFixture.ts:4]: please update file/function/variable naming

File: apps/smart-contracts/core/test/ManagerWithdrawHook.test.ts [hunk 7]
```diff
@@ -114,39 +114,15 @@ describe('=> ManagerWithdrawHook', () => {
...
+    it('succeeds if role holder', async () => {
+      expect(
+        await managerWithdrawHook.hasRole(
+          await managerWithdrawHook.SET_COLLATERAL_ROLE(),
+          deployer.address
+        )
+      ).to.eq(true)
 
-      await expect(tx).to.emit(managerWithdrawHook, 'CollateralChange').withArgs(collateral.address)
+      await expect(managerWithdrawHook.connect(deployer).setCollateral(collateral.address))
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/ManagerWithdrawHook.test.ts:125]: this doesn't need to be inside `expect`

File: apps/smart-contracts/core/test/UsesCollateral.test.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  let deployer: SignerWithAddress
+  let user: SignerWithAddress
+  let collateral: SignerWithAddress
+  beforeEach(async () => {
+    ;[deployer, user, collateral] = await ethers.getSigners()
+    usesCollateral = await usesCollateralFixture()
+  })
+
+  describe('# initialize', () => {
+    it('doesnt have collateral set', async () => {
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/UsesCollateral.test.ts:19]:
```suggestion
    it('sets collateral to zero address', async () => {
```

**Review by XavierEkkel — APPROVED** [review]
LGTM after comments

**Pull request merged**